### ISSUE TITLE
Correctly support the postgres "serializable" and "repeatable read" isolation levels

### ISSUE THREAD

**Issue by lorenzo** [hasura/graphql-engine#7498]
While it is possible to instruct Hasura to always use the `SERIALIZABLE` isolation level, Hasura does not provide the logic that this isolation level requires to function properly.

From the [postgres documentation](https://www.postgresql.org/docs/13/transaction-iso.html):

> The Serializable isolation level provides the strictest transaction isolation. This level emulates serial transaction execution for all committed transactions; as if transactions had been executed one after another, serially, rather than concurrently. However, like the Repeatable Read level, applications using this level **must be prepared to retry transactions due to serialization failures**

(emphasis mine)

What this means in practice is a rather simple change in the server logic: Whenever the `40001` error code is returned by Postgres, execute the same transaction until it succeeds or errors out with a different code. If you are familiar with Haskell's STM, this concept should already be familiar.

The lack of this retry behavior makes it impractical to use `repeatable read` or `serializable` in real applications with concurrency, as the error will bubble up the clients who will have no clue if the mutation should be attempted again or not.

**Comment by abdullah2993** [comment]
yes, +1 for this one we use `SERIALIZABLE` with `HASURA_GRAPHQL_TX_ISOLATION` under load it just starts throwing `40001`

**Comment by tirumaraiselvan** [comment]
> The lack of this retry behavior makes it impractical to use repeatable read or serializable in real applications with concurrency, as the error will bubble up the clients who will have no clue if the mutation should be attempted again or not.

We should atleast be surfacing serialization errors to the client through the `code` in the error. Will that work?

**Comment by lorenzo** [comment]
The most desirable solution is to handle it on the backend so that you can get a better chance of your mutation going through or receive an actual error. The two downsides of having to manage this on the clients are:

* Every client needs to do it, and you end up repeating a lot of code.
* When exposing the graph publicly, you need to document that mutations *required* being retried. People may miss this important part.
* you risk having lots of clients sending retries with almost equal chances of them failing, creating artificial load in the system
